crypto/src/main/java/org/web3j/crypto/StructuredDataEncoder.java

```diff
@@ -0,0 +1,448 @@
+                pureDimensions.add((Integer) depthDimensionPair.getSecond());
+            }
+            depthDimensionsMap.put((Integer) entry.getKey(), pureDimensions);
+        }
+
+        ArrayList<Integer> dimensions = new ArrayList<>();
+        for (Map.Entry<Integer, List<Integer>> entry : depthDimensionsMap.entrySet()) {
+            Set<Integer> setOfDimensionsInParticularDepth = new TreeSet<>(entry.getValue());
+            if (setOfDimensionsInParticularDepth.size() != 1) {
+                throw new ValidationException(
+                        String.format(
+                                "Depth %d of array data has more than one dimensions",
+                                entry.getKey()
+                        )
+                );
+            }
+            dimensions.add(setOfDimensionsInParticularDepth.stream().findFirst().get());
+        }
+
+        return dimensions;
+    }
+
+    public ArrayList<Object> flattenMultidimensionalArray(Object data) {
+        if (!ArrayList.class.isInstance(data)) {
+            return new ArrayList<Object>() {
+                {
+                    add(data);
+                }
+            };
+        }
+
+        ArrayList<Object> flattenedArray = new ArrayList<>();
+        for (Object arrayItem: (ArrayList) data) {
+            for (Object otherArrayItem: flattenMultidimensionalArray(arrayItem)) {
+                flattenedArray.add(otherArrayItem);
+            }
+        }
+
+        return flattenedArray;
+    }
+
+    public byte[] encodeData(
+            String primaryType,
+            LinkedHashMap<String, Object> data) throws ValidationException {
+        HashMap<String, List<StructuredData.Entry>> types = jsonMessageObject.getTypes();
+
+        List<String> encTypes = new ArrayList<>();
+        List<Object> encValues = new ArrayList<>();
+
+        // Add typehash
+        encTypes.add("bytes32");
+        encValues.add(typeHash(primaryType));
+
+        // Add field contents
+        for (StructuredData.Entry field: types.get(primaryType)) {
+            Object value = data.get(field.getName());
+
+            if (field.getType().equals("string")) {
+                encTypes.add("bytes32");
+                byte[] hashedValue = Numeric.hexStringToByteArray(sha3String((String) value));
+                encValues.add(hashedValue);
+            } else if (field.getType().equals("bytes")) {
+                encTypes.add("bytes32");
+                byte[] hashedValue = sha3((byte[]) value);
+                encValues.add(hashedValue);
+            } else if (types.containsKey(field.getType())) {
+                // User Defined Type
+                byte[] hashedValue = sha3(
+                        encodeData(field.getType(), (LinkedHashMap<String, Object>) value)
+                );
+                encTypes.add("bytes32");
+                encValues.add(hashedValue);
+            } else if (Pattern.matches(arrayTypeRegex, field.getType())) {
```

Comment:
Are we better off pulling this out into a field so the pattern doesn't have to get compiled each time this function is run?